api/registrations/serializers.py

```diff
@@ -567,13 +569,55 @@ def _find_orphan_files(self, registering, draft):
         from website.archiver.utils import find_selected_files
         files = find_selected_files(draft.registration_schema, draft.registration_metadata)
         orphan_files = []
-        for _, value in files.items():
+        for key, value in files.items():
             if 'extra' in value:
                 for file_metadata in value['extra']:
-                    if file_metadata['nodeId'] not in registering:
+                    if not self._is_attached_file_valid(file_metadata, registering):
                         orphan_files.append(file_metadata)
+
         return orphan_files
 
+    def _is_attached_file_valid(self, file_metadata, registering):
+        """
+        Validation of file information on registration_metadata.  Theoretically, the file information
+        on registration_responses does not have to be valid, so we enforce their accuracy here,
+        to ensure file links load properly.
+
+        Verifying that nodeId in the file_metadata is one of the files we're registering. Verify
+        that selectedFileName is the name of a file on the node.  Verify that the sha256 matches
+        a version on that file.
+
+        :param file_metadata - under "registration_metadata"
+        :param registering - node ids you are registering
+        :return boolean
+        """
+
+        node_id = file_metadata.get('nodeId')
+        if node_id not in registering:
+            return False
+
+        node = AbstractNode.load(node_id)
+        if not node:
+            # node in registration_metadata doesn't exist
+            return False
+
+        specified_sha = file_metadata.get('sha256', '')
+
+        file = node.files.filter(name=file_metadata.get('selectedFileName')).first()
+        if not file:
+            # file with this name does not exist on the node
+            return False
+
+        match = False
+        for version in file.versions.all():
+            if specified_sha == version.metadata.get('sha256'):
+                match = True
+
+        if not match:
+            # Specified sha256 does not match a version on the specified file
+            return False
+
+        return True
```

Comment:
hmm, this is going to need lots of testing, if we're validating file information before registering.

Note that I'm still using fields on `registration_metadata` here because the archiving process is still looking at these fields.  I am attempting to use fields that should always exist, regardless of whether `registration_metadata` was written to directly, or it was built from `registration_responses`. v 

A question I do have is what if the file is renamed after the file is linked in the registration responses?

Reply:
Hmm, if the file is renamed after it's linked to the draft but before the final registration... I think I'd expect `file_name` to be updated with the actual name when the registration is created? But it might make more sense to think of it as a linked file _version_, so we'd use the name and hash from when the user selected it in the draft.

Thinking about the archiving process... does the archiver update the registration's `registered_meta` (or the draft's `registered_metadata`) with the view URL of the archived file? Do we need to make it also update `registered_responses`?

Reply:
The archiver updates the registrations `registered_meta` with the viewURL of the archived file.     https://github.com/CenterForOpenScience/osf.io/blob/develop/website/archiver/utils.py#L310

`registration_metadata/registered_meta` is still pretty integral to archiver, that code hasn't been touched.  You can update `registered_responses` too, but I wouldn't rely on `registered_responses` alone unless you're updating archiver code at this point.  And even still, we need to maintain both workflows for now?

Reply:
Yeah let's not redo archiver just yet. I think we'll want to leave the logic that uses `registered_meta` to locate the linked files/hashes, check that they're present and valid, and update `registered_meta` with correct URLs; then just add a bit that *also* updates `registration_responses`.